### ISSUE TITLE
unposition all modules

### ISSUE THREAD

**Issue by hjkirkwood** [European-XFEL/karabo_data#139]
Dear experts,

I have created a mask based on a pixels distance from the center of the assembled agipd image and would like to un-assemble it and go back to the stacked data shape of (16,512,128) based on the geometry file. 
Could that be implemented? I had a look at the code but got a bit lost.

The idea is that I could apply the following and then undo the assemble operation:

```
geom = AGIPD_1MGeometry.from_crystfel_geom(geom_file)

assem_im, center = geom.position_modules_fast(np.ones((16,512,128)))

xg, yg = np.mgrid[0:assem_im.shape[0], 0:assem_im.shape[1]]
radius = np.hypot(xg - center[0],yg - center[1])

rmin = 450
rmax = 550
radius[( (radius<rmin) | (radius>rmax) )]=0
radial_mask = (radius>0)
```

**Comment by takluyver** [comment]
This ties in with something we were discussing yesterday - I think the geometry classes should have methods to convert between locations in data space (module number, slow scan, fast scan) and real space (x, y, z).

**Comment by hjkirkwood** [comment]
Yes this would also be very useful too, I assume you mean to convert from data space to real space and vice versa also. I'd be very interested to have this functionality! Thanks

**Comment by takluyver** [comment]
I might wait until #129 is merged to tackle this, because I think it might be able to share some of the code.

**Comment by takluyver** [comment]
@hjkirkwood I think I've worked out an efficient way to do this kind of thing: #142 would add a `geom.get_pixel_positions()` method giving you an array of (x, y, z) positions for each pixel, in the same shape as one frame of data. So writing a radial mask should be something like this (not tested):

```python
pixelpos = geom.get_pixel_positions()  # shape: (16, 512, 128, 3)

radius = np.sqrt(np.sum(pixelpos ** 2, axis=-1))  # shape: (16, 512, 128)
radius_mask = (radius > 450) & (radius < 550)  # shape: (16, 512, 128)
```

(This is actually calculating the radius in 3D, i.e. it masks all points between two spheres. If the detector is flat it should be the same; or you can discard the Z coordinates to do it in 2D)

Let me know if the API works for you and if there are any changes you want. Once we've agreed how it should work, I'll add docs and tests.

**Comment by antarcticrainforest** [comment]
What happens here if you wanted to do more complicated operations or masking certain ROI's in the 2D plane?

**Comment by takluyver** [comment]
You'd need to express the mask as conditions which could be checked against the pixel arrays. E.g. selecting a rectangular ROI would be something like this:

```python
px = pixelpos[..., 0]
py = pixelpos[..., 1]

roi = (px > 100) & (px < 200) & (py > 50) & (py < 250)
```

This reminds me that there's an API question. I've currently put the (x, y, z) coordinates as the last axis in the array, but it could also be the first. It may have some potential implications for processing efficiency, but I'm not confident without timing it, and it probably doesn't need to be done in a tight loop anyway, so I don't think raw speed matters that much anyway.

Maybe I should make it a labelled array so the axis order is less important.

**Comment by takluyver** [comment]
@hjkirkwood : just a reminder, further progress on this is waiting for your input on the API described in the comments above. I don't want to add something that's not useful, so I'm hoping you'll confirm whether it would let you do what you want.

I'm happy to answer any questions about the idea. :-)

**Comment by hjkirkwood** [comment]
Hey,

I'm using this now and currently tossing up a few things. I used that "robust peak finder" we played around with a while ago, I ran it on each module individually (assembling the image is not neccesary and this is much simpler). Now I have a table of hits where each row contains the info for each peak (location within the module, area, intensity etc.) as well as the trainId, pulseId and module number. Next I convert this to the lab/detector coordinates using 'get_pixel_positions()' and some interpolation and this works very nicely, thanks! This does exactly as needed.

My only question is how is the number calculated - is it just the indices of the assembled array minus the center positions? or is it the pixel center (+/- 0.5)

**Comment by takluyver** [comment]
I could add a method to directly convert data coordinates (module num, slow scan, fast scan) to physical coordinates (x, y, z), if that would save you some interpolation. It could be either instead of the proposed `get_pixel_positions()` or in addition to that.

The method in the PR defaults to calculating the centre position of each pixel (assuming I've implemented it right). If you pass `centre=False`, it will instead give the position of the pixel corner nearest the start. i.e. if you have a module positioned neatly at the origin:

- `centre=False` puts the first pixel at (0, 0, 0)
- `centre=True` (default) puts the first pixel at (0.5, 0.5, 0)

This also reminds me: do you have any preference on the units for physical coordinates? I'm considering moving most of the internals of the geometry module to use metres rather than 'pixel units' (#160). But obviously it's easy enough to convert back for public APIs.

**Comment by hjkirkwood** [comment]
Ah yes I see now, it helps if I read the documentation ;)

Yes, an additional method would be useful I just did nearest neighbor interpolation to convert subpixel location from module to lab frame (i.e. (module num, slow scan,  fast scan) should take floats as input for slow and fast scan).  I'd suggest keeping the get_pixel_positions() as there are still many different use cases for this.

For the units, my preference is pixels, then I convert to physical units later in the analysis. I find it is easier to keep track of and for sanity checking at this stage (during hit finding etc.)

**Comment by takluyver** [comment]
Thanks for the feedback. PR #142 now also includes `data_coords_to_physical()` which should do what we discussed. The data coordinates are taken to be from the corner of the module, so 0.5, 0.5 is the centre of the first pixel.

Results at tile boundaries will differ slightly from interpolating, because this code will assign each set of coordinates to one tile. There's still no special handling for the double-width pixels at the edges of tiles.

This currently only works for AGIPD, but I imagine that's what you're interested in. It should be feasible for LPD but fiddlier to select the tile. It will be messy to do accurately for DSSC.

**Comment by takluyver** [comment]
PS: I've endeavoured to avoid iterating over the coordinates in Python, with the idea that it should efficiently convert many data coordinates (passed as numpy arrays). But I haven't done any performance testing, so there may well be room for improvement.

**Comment by takluyver** [comment]
@hjkirkwood I think #142 is ready to merge if these APIs work for you:

```
pixelpos = geom.get_pixel_positions()
x, y, z = pixelpos[module_no, slow_scan, fast_scan]

# And:

coords = geom.data_coords_to_physical(module_no, slow_scan, fast_scan) # Numpy arrays
x = coords[..., 0]
```

The first one gives you a position (by default the centre) for each pixel in the detector, to conveniently select them for masks. The second lets you transform arrays of data coordinates (e.g. from robust peak finder) into physical positions.

In both cases, the dimension for the x/y/z parts of the coordinates is currently the innermost, but it could easily be the outermost if you prefer. That would mean you could use unpacking `px, py, pz = arr` to get three separate arrays for the three dimensions, rather than `px = arr[..., 0]` etc.

**Comment by takluyver** [comment]
I've now switched both APIs to return coordinates in metres, as requested by @fangohr. I also kind of like that using metres gives a very obvious distinction between physical coordinates (<1 metre) and data coordinates (>>1 pixel).

But it's still easy to get the results in pixel units if you prefer that - just divide by the pixel size:

```
pixelpos = geom.get_pixel_positions() / geom.pixel_size
coords = geom.data_coords_to_physical(module_no, slow_scan, fast_scan) / geom.pixel_size
```

**Comment by takluyver** [comment]
Closed by #142
